File: core/src/main/java/org/seppiko/wagashi/core/controllers/AccountController.java
package org.seppiko.wagashi.core.controllers;

import org.seppiko.commons.utils.StringUtil;
import org.seppiko.wagashi.commons.utils.JsonUtil;
import org.seppiko.wagashi.commons.utils.JwtUtil;
import org.seppiko.wagashi.core.models.ResponseMessageEntity;
import org.seppiko.wagashi.core.models.ResponseTokenEntity;
import org.seppiko.wagashi.core.services.UserPrincipal;
import org.seppiko.wagashi.core.services.UserService;
import org.seppiko.wagashi.core.utils.ResponseUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RestController;

/**
 * @author Leonard Woo
 */
@RestController
public class AccountController {

  @Autowired
  private UserService service;

  @RequestMapping(value = "/login", method = RequestMethod.POST)
  public ResponseEntity<byte[]> loginContentHandleExecution(@RequestBody(required = false) String requestBody) {
    if (requestBody == null || StringUtil.hasLength(requestBody)) {
      return ResponseUtil.sendJson(400, new ResponseMessageEntity<>(400, "failure"));
    }

    var login = JsonUtil.fromJsonObject(requestBody);
    if (login.isEmpty()) {
      return ResponseUtil.sendJson(200, new ResponseMessageEntity<>(400, "failure"));
    }

    UserDetails userDetails = service.loadUserByUsername(login.get("username").asText());
    if (!login.get("password").asText().equals(userDetails.getPassword())) {
      return ResponseUtil.sendJson(200, new ResponseMessageEntity<>(401, "failure"));
    }

    String jwt = JwtUtil.jwtGenerator(userDetails.getUsername());
    ResponseTokenEntity respToken = new ResponseTokenEntity(jwt, JwtUtil.getExpires());

    return ResponseUtil.sendJson(200, new ResponseMessageEntity<>(200, "ok", respToken));
  }

  @RequestMapping(value = "/logout")
  public ResponseEntity<byte[]> logoutContentHandleExecution() {
    return ResponseUtil.sendJson(200, new ResponseMessageEntity<>(200, "ok"));
  }

}

File: core/src/main/kotlin/org/seppiko/wagashi/core/controllers/AccountController.kt
/*
 * Copyright 2022 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.seppiko.wagashi.core.controllers

import org.seppiko.wagashi.core.models.ResponseMessage
import org.seppiko.wagashi.core.services.UserService
import org.seppiko.wagashi.core.utils.ResponseUtil.sendJson
import org.seppiko.wagashi.commons.utils.JsonUtil
import org.seppiko.wagashi.commons.utils.JwtUtil
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod
import org.springframework.web.bind.annotation.RestController
import java.util.stream.Collectors


/**
 *
 * @author Leonard Woo
 */
@RestController
class AccountController {

  @Autowired
  private lateinit var service: UserService


  @RequestMapping(value = ["/login"], method = [RequestMethod.POST])
  fun ContentHandleExecution(
    @RequestBody(required = false) requestBody: String?
  ): ResponseEntity<ByteArray?>? {

    if (requestBody.isNullOrBlank()) {
      return sendJson(400, ResponseMessage<Any>(400, "failure"))
    }

    val login = JsonUtil.fromJsonObject(requestBody)
    val user = service.loadUserByUsername(login!!.get("username").asText())
    if (!login.get("password").asText().equals(user.password)) {
      return sendJson(200, ResponseMessage<Any>(401, "failure"))
    }

    val jwt = JwtUtil.generator(user.username)

//    val roles: List<String> = user.authorities.stream()
//      .map { item -> item.authority }
//      .collect(Collectors.toList())

    val map = mapOf(
      "access_token" to jwt,
      "token_type" to "Bearer",
      "expires_in" to 3600,
    )

    return sendJson(200, ResponseMessage<Any>(200, "ok", map))
  }

}